File: packages/google_mlkit_translation/android/src/main/java/com/google_mlkit_translation/TextTranslator.java
package com.google_mlkit_translation;

import androidx.annotation.NonNull;

import com.google.mlkit.nl.translate.TranslateRemoteModel;
import com.google.mlkit.nl.translate.Translation;
import com.google.mlkit.nl.translate.Translator;
import com.google.mlkit.nl.translate.TranslatorOptions;
import com.google_mlkit_commons.GenericModelManager;

import java.util.HashMap;
import java.util.Map;

public class TextTranslator implements Pigeon.OnDeviceTranslatorApi {
    private final Map<String, Translator> instances = new HashMap<>();
    private final GenericModelManager genericModelManager = new GenericModelManager();

    @Override
    public void translateText(
        @NonNull Pigeon.TranslateRequest request, 
        @NonNull Pigeon.Result<String> result
        ) {
        String id = request.getId();
        Translator onDeviceTranslator = instances.get(id);

        if (onDeviceTranslator == null) {
            TranslatorOptions options = new TranslatorOptions.Builder()
                    .setSourceLanguage(request.getSourceLanguage())
                    .setTargetLanguage(request.getTargetLanguage())
                    .build();
            onDeviceTranslator = Translation.getClient(options);
            instances.put(id, onDeviceTranslator);
        }
        final Translator translator = onDeviceTranslator;
        final String text = request.getText();

        translator.downloadModelIfNeeded()
                .addOnSuccessListener(
                        (OnSuccessListener) -> {
                            // Model downloaded successfully. Okay to start translating.
                            translator.translate(text)
                                    .addOnSuccessListener(result::success)
                                    .addOnFailureListener(
                                            e -> result.error(new Exception("Error translating: " + e.getMessage())));
                        })
                .addOnFailureListener(
                        e -> {
                            // Model could not be downloaded, or there was another internal error.
                            result.error(new Exception("Error building translator. Either source or target models are not downloaded: " + e.getMessage()));
                        });
    }

    @Override
    public void closeTranslator(@NonNull Pigeon.CloseTranslatorRequest request) {
        String id = request.getId();
        Translator translator = instances.get(id);
        if (translator == null) return;
        translator.close();
        instances.remove(id);
    }

    @Override
    public void manageModel(@NonNull Pigeon.ModelManagementRequest request, @NonNull Pigeon.Result<Pigeon.ModelManagementResponse> result) {
//        TranslateRemoteModel model = new TranslateRemoteModel.Builder(request.getModel()).build();
//        String task = request.getTask();
//        genericModelManager.manageModel(model, request, result);
    }

}

File: packages/google_mlkit_translation/android/src/main/kotlin/com/google_mlkit_translation/TextTranslator.kt
package com.google_mlkit_translation

import com.google.mlkit.nl.translate.TranslateRemoteModel
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.Translator
import com.google.mlkit.nl.translate.TranslatorOptions
import com.google_mlkit_commons.GenericModelManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class TextTranslator : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, Translator>()
    private val genericModelManager = GenericModelManager()

    companion object {
        private const val START = "nlp#startLanguageTranslator"
        private const val CLOSE = "nlp#closeLanguageTranslator"
        private const val MANAGE = "nlp#manageLanguageModelModels"
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                translateText(call, result)
            }

            CLOSE -> {
                closeDetector(call)
                result.success(null)
            }

            MANAGE -> {
                manageModel(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun translateText(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val text = call.argument<String>("text") ?: return
        val id = call.argument<String>("id") ?: return

        val translator =
            instances.getOrPut(id) {
                val sourceLanguage = call.argument<String>("source") ?: return
                val targetLanguage = call.argument<String>("target") ?: return
                Translation.getClient(
                    TranslatorOptions
                        .Builder()
                        .setSourceLanguage(sourceLanguage)
                        .setTargetLanguage(targetLanguage)
                        .build(),
                )
            }

        translator
            .downloadModelIfNeeded()
            .addOnSuccessListener {
                translator
                    .translate(text)
                    .addOnSuccessListener { result.success(it) }
                    .addOnFailureListener { e -> result.error("error translating", e.toString(), null) }
            }.addOnFailureListener {
                result.error("Error building translator", "Either source or target models not downloaded", null)
            }
    }

    private fun closeDetector(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)?.close()
    }

    private fun manageModel(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val model = TranslateRemoteModel.Builder(call.argument<String>("model").toString()).build()
        genericModelManager.manageModel(model, call, result)
    }
}
